### ISSUE TITLE
[bug] PT1.10.0 DLC error loading torchvision.ops.nms()

### ISSUE THREAD

**Issue by sitecao** [aws/deep-learning-containers#1631]
Checklist
- [x] I've prepended issue tag with type of change: [bug]
- [x] (If applicable) I've attached the script to reproduce the bug
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] I'm using an existing DLC image listed here: https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-images.html
- [ ] I've built my own container based off DLC (and I've attached the code used to build my own image)

***Concise Description:*** Both the E3 and SM PT1.10.0 DLC encounter error when trying to use this non maximum suppression op. This is not seen in other containers I tried (e.g. PT1.9.1)

***DLC image/dockerfile:*** 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:1.10.0-gpu-py38-cu113-ubuntu20.04-sagemaker

***Current behavior:*** Shows below error (scroll to bottom to see the simple script I use to reproduce):
Traceback (most recent call last):
  File "r.py", line 12, in <module>
    torchvision.ops.nms(x, y, t)
  File "/opt/conda/lib/python3.8/site-packages/torchvision/ops/boxes.py", line 34, in nms
    _assert_has_ops()
  File "/opt/conda/lib/python3.8/site-packages/torchvision/extension.py", line 26, in _assert_has_ops
    raise RuntimeError(
RuntimeError: Couldn't load custom C++ ops. This can happen if your PyTorch and torchvision versions are incompatible, or if you had errors while compiling torchvision from source. For further information on the compatible versions, check https://github.com/pytorch/vision#installation for the compatibility matrix. Please check your PyTorch version with torch.__version__ and your torchvision version with torchvision.__version__ and verify if they are compatible, and if not please reinstall torchvision so that it matches your PyTorch install.

***Expected behavior:*** Not error out.

***Additional context:*** One hack on my side that worked is to force reinstall PyTorch of the same version. Both PyTorch and torchvision will show the exact same version but it will work after reinstalling although I am not quite sure why:

**Command to force reinstall:**
pip3 install torch==1.10.0+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html --force-reinstall

**Python script to reproduce:**

import torch
import torchvision

x = torch.tensor([[90614.00000, 90225.75000, 90632.00000, 90236.75000]]).cuda()
y = torch.tensor([0.00209]).cuda()
t = 0.6

print(torchvision.ops.nms(x, y, t))

**Comment by krzim-aws** [comment]
I've run into the same issue while trying to run Detectron2 in an AWS DLC. I've tracked the issue down to what is likely a build issue with the AWS EI PyTorch binary used in the DLC. 

The above error message isn't all that useful in determining the underlying cause but if you attempt to load the `torchvision` extension library in a similar manner to this line [here](https://github.com/pytorch/vision/blob/e92f1195f4410d36133c22efac7868ee61cb8bbd/torchvision/extension.py#L20), you will see that `libtorch_cuda_cu.so` and `libtorch_cuda_cpp.so` cannot be located. 

In `torch 1.9.0`  and PR https://github.com/pytorch/pytorch/pull/52503 the `BUILD_SPLIT_CUDA` flag was automatically set which splits the `libtorch_cuda` library into `libtorch_cuda_cu` and `libtorch_cuda_cpp` as seen [here](https://github.com/pytorch/pytorch/blob/abb55c53b3faf41fb1b2e9a69a1ca7b49c3e5b1b/torch/utils/cpp_extension.py#L997).

I'm guessing that the EI PyTorch binaries (like the one used [here](https://github.com/aws/deep-learning-containers/blob/51b910fca041e8ef380099fab60fe4f8c5dafda3/pytorch/inference/docker/1.10/py3/cu113/Dockerfile.e3.gpu#L10)) were compiled without this flag and the AWS torchvision binaries (like the one use [here](https://github.com/aws/deep-learning-containers/blob/51b910fca041e8ef380099fab60fe4f8c5dafda3/pytorch/inference/docker/1.10/py3/cu113/Dockerfile.e3.gpu#L11)) were compiled against a `torch` install that was compiled with the flag.

As the build pipelines appear not to be opensource, it would require a change to AWS's internal build pipelines for one or both of these binaries to resolve this issue. I'm currently attempting to build `torchvision` from source in the `763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-inference:1.10.0-gpu-py38` to see if this will resolve the issue.

It might also be necessary to add  `/opt/conda/lib/python3.8/site-packages/torch/lib` to `LD_LIBRARY_PATH` in the image.

**Comment by krzim-aws** [comment]
As a workaround, I suggest just building `torchvision` from source in the DLC with something like the following:

```
FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:1.10.0-gpu-py38

ENV FORCE_CUDA=1
ARG TORCHVISION_VERSION=0.11.1

RUN wget -q https://github.com/pytorch/vision/archive/refs/tags/v${TORCHVISION_VERSION}.tar.gz \ 
    && mkdir -p vision \
    && tar -xzf v${TORCHVISION_VERSION}.tar.gz -C vision --strip-components 1 \
    && cd vision \
    && python setup.py install \
    && rm -rf /tmp/vision
```

This worked for me and allowed me to build and run Detectron2 in a SM DLC. I've also tested this with the `torch==1.9.0` DLC and the matching release of `torchvision==0.10.0`.

The AWS team could likely resolve this by building the `torchvision` binary used in the DLCs against the `torch` binary used in those same DLCs. My guess is that this might not currently be the case per my previous comment.

**Comment by tejaschumbalkar** [comment]
This issue is resolved in latest PyTorch DLC v2.2
`763104351884.dkr.ecr.us-west-2.amazonaws.com/pytorch-training:2.2.0-gpu-py310-cu121-ubuntu20.04-sagemaker-v1.0 `

Can you try using the DLC and confirm the results?

**Comment by arjkesh** [comment]
Closing issue due to lack of response. Please feel free to reopen if this issue persists.
